pkg/featurechangestopper/featurechangestopper.go

```diff
@@ -0,0 +1,140 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configinformersv1 "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+// TechPreviewChangeStopper calls stop when the value of the featuregate changes from TechPreviewNoUpgrade to anything else
+// or from anything to TechPreviewNoUpgrade.
+type TechPreviewChangeStopper struct {
+	startingTechPreviewState bool
+
+	featureGateLister configlistersv1.FeatureGateLister
+	cacheSynced       []cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+}
+
+// New returns a new cluster version operator.
+func New(
+	startingTechPreviewState bool,
+	featureGateInformer configinformersv1.FeatureGateInformer,
+) *TechPreviewChangeStopper {
+	c := &TechPreviewChangeStopper{
+		startingTechPreviewState: startingTechPreviewState,
+		featureGateLister:        featureGateInformer.Lister(),
+		cacheSynced:              []cache.InformerSynced{featureGateInformer.Informer().HasSynced},
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "feature-gate-stopper"),
+	}
+
+	featureGateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			c.queue.Add("cluster")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+	})
+
+	return c
+}
+
+func (c *TechPreviewChangeStopper) syncHandler(ctx context.Context, shutdownFn context.CancelFunc) error {
+	featureGates, err := c.featureGateLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+	techPreviewNowSet := featureGates.Spec.FeatureSet == configv1.TechPreviewNoUpgrade
+	if techPreviewNowSet != c.startingTechPreviewState {
+		klog.Infof("TechPreviewNoUpgrade status changed from %v to %v, shutting down.", c.startingTechPreviewState, techPreviewNowSet)
+		if shutdownFn != nil {
+			shutdownFn()
+		}
+		return nil
+	}
+	return nil
+}
+
+func (c *TechPreviewChangeStopper) Run(runContext context.Context, shutdownFn context.CancelFunc) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
+	// make sure the work queue is shutdown which will trigger workers to end
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting stop-on-techpreview-change controller")
+
+	// wait for your secondary caches to fill before starting your work
+	if !cache.WaitForCacheSync(runContext.Done(), c.cacheSynced...) {
+		return
+	}
+
+	// runWorker will loop until "something bad" happens.  The .Until will
+	// then rekick the worker after one second
+	go wait.UntilWithContext(runContext, func(ctx context.Context) {
```

Comment:
why launch a goroutine here?  The existing gorountine is just off to wait on `runContext` and then log the shutdown.  Can't you:

```golang
  wait...whatever...

  klog.Infof("Shutting down stop-on-techpreview-change controller")
}

Reply:
> why launch a goroutine here? The existing gorountine is just off to wait on runContext and then log the shutdown. Can't you:

so that they stack neatly when you have more than one.  this allows parallel structures across every controller.  See the source link above.

Reply:
Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.